feat: add support for dark mode images and semantic colors

ewanharris opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27126

This introduces support for iOS 13's dark mode. There's two parts to the PR

* b3ce21fc842ffb6fe137748e1317ec7846fa31ef - Adds support for adding a dark mode variant of an image by adding the `-dark` suffix to a an image name (but before the retina `@2x/@3x`).
* f8bd718ad6fb59017d959bcdeacbb6550a9d24a8 - Adds support for the iOS semantic colors, but allows reusability cross platform by loading the semantic.colors.json file on Android/Windows and allowing developers to set a preferred color type via `Ti.UI.semanticColorType`

https://github.com/ewanharris/darkmode-example can be used as a test/example

ewanharris committed:
feat(ios): add support for specifying dark mode images

Implements part of TIMOB-27126

ewanharris committed:
feat: add support for semantic colors

This adds a cross platfom method for loading semanitc colors, on iOS 11+ we will use the native Ti.UI.iOS.fetchSemanticColor to load the right color, in all other cases we use the Ti.UI.semanticColorType and the provided json file to obtain the correct value

Fixes TIMOB-27126

build commented:
<!--
  2 failure:  :microscope: Ther..., :disappointed_rel...
  1 warning:  :mag: Can't find ...
  2 messages
  
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/ewanharris/titanium_mobile/blob/TIMOB-27126/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  :mag: Can't find junit reports at `./junit.*.xml`, skipping generating JUnit Report.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@8.2.0 test /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-11097
> npm run ios-sanity-check && grunt


> titanium-mobile@8.2.0 ios-sanity-check /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-11097
> ./build/scons check-ios-toplevel

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-11097/cli/lib/tasks/process-js-task.js
  157:12  warning  Avoid calling back inside of a promise  promise/no-callback-in-promise
  164:5   warning  Avoid calling back inside of a promise  promise/no-callback-in-promise

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-11097/iphone/cli/hooks/frameworks.js
   60:5   warning  Expected catch() or return                  promise/catch-or-return
   60:34  warning  Avoid calling back inside of a promise      promise/no-callback-in-promise
  425:5   warning  Each then() should return a value or throw  promise/always-return

✖ 5 problems (0 errors, 5 warnings)


Running "checkFormat:ios" (checkFormat) task

Running "checkFormat:android" (checkFormat) task

Running "validate:docs" task
Warning: Docs validation failed with exit code: 1
Output: [ERROR] ./Titanium/UI/UI.yml
Titanium.UI: found 1 error(s)!
	properties
		semanticColorType
			Required property "type" not found

[ERROR] Found 1 error(s)!
 Use --force to continue.

Aborted due to warnings.
npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 064739e4ec38be42ea3b12cedfafa797f3772b5f
</p>

ewanharris committed:
fix: correct check for iOS namespace

ewanharris commented on iphone/Classes/TiUIiOSProxy.m:
```diff
@@ -841,5 +841,16 @@ - (id)createWebViewProcessPool:(id)args
+{
+  ENSURE_SINGLE_ARG(color, NSString);
+
+  if ([TiUtils isIOSVersionOrGreater:@"11.0"]) {
```
I think we'll want to merge this after https://github.com/appcelerator/titanium_mobile/pull/11091 so I can wrap this in a check?

ewanharris commented on iphone/Classes/TiUIiOSProxy.m:
```diff
@@ -841,5 +841,16 @@ - (id)createWebViewProcessPool:(id)args
+  if ([TiUtils isIOSVersionOrGreater:@"11.0"]) {
+    return [[TiColor alloc] initWithColor:[UIColor colorNamed:color] name:nil];
+  } else {
+    return [[TiColor alloc] initWithColor:UIColor.blackColor name:@"black"];
```
While this is undocumented, it's still callable by a user. So I kept the fallback to black like in the darkmode plugin/module by Hans.

vijaysingh-axway commented on build/lib/packager.js:
```diff
@@ -275,10 +275,12 @@ class Packager {
 				babel(babelOptions)
 			],
-			external: [ './app', 'com.appcelerator.aca' ]
+			external: [ './app', 'com.appcelerator.aca', '/semantic.colors.json' ]
```
@ewanharris I am getting error, if I run titanium SDK using Xcode.
`[ERROR] Script Error {
    column = 46;
    line = 5;
    message = "Couldn't find module: ../../../semantic.colors.json for architecture: x86_64";`

ewanharris replied:
I forgot to copy the rollup options to the `scons-xcode-project-build` script. Will fix

ewanharris replied:
Fixed in 8e732a37a3c1161f7ade223f85fc13734fb5272d

vijaysingh-axway commented on common/Resources/ti.internal/extensions/ti/ti.ui.js:
```diff
@@ -0,0 +1,46 @@
+		osVersion = parseInt(Ti.Platform.version.split('.')[0]);
+	}
+
+	if (Ti.App.iOS && osVersion >= 11) {
```
If I run on 11 < iOS < 13, color is not getting changed for dark and light mode set from Ti.UI.semanticColorType. Here osVersion check should be 
`if (Ti.App.iOS && osVersion >= 13) { }` 
.

ewanharris replied:
So the API we use [colorNamed](https://developer.apple.com/documentation/uikit/uicolor/2877380-colornamed?language=objc) is available from iOS 11 onwards, I figured it was always better to fall back to the native API if possible rather than our mocked dark mode, but yeah it only returns light colors. I'll make this check 13+ so we can handle that

ewanharris replied:
Updated

jquick-axway commented on common/Resources/ti.internal/extensions/ti/ti.ui.js:
```diff
@@ -0,0 +1,46 @@
+	} else {
+		if (!colorset) {
+			try {
+				colorset = require('/semantic.colors.json'); // eslint-disable-line import/no-absolute-path
```
If we load the JSON file like we do in the link below, then we won't run into a `require()` error when running via Android Studio, Xcode, or Visual Studio. This way the native dev team here can debug via our IDEs without issue.
https://github.com/appcelerator/titanium_mobile/blob/master/common/Resources/ti.internal/bootstrap.loader.js#L31

ewanharris replied:
iOS works fine after 08a592fa7ae81e9db39fc0a68c4f25e0cd33c4c0, I'm also able to build from Android Studio just fine after updating the `ti.main.js` file in TitaniumTest to one from this PR. I believe as long as rollup is configured correctly it shouldn't be an issue

ewanharris committed:
build: apply rollup configuration to xcode project build

ewanharris committed:
fix: fall back to json file when using below ios 13

vijaysingh-axway commented on apidoc/Titanium/UI/UI.yml:
```diff
@@ -2309,6 +2319,23 @@ properties:
+    since: "8.2.0"
+
+  - name: SEMANTIC_COLOR_TYPE_LIGHT
+    summary: Return the dark value from the applications colorset.
```
"dark" should be "light" here.

ewanharris replied:
Fixed

vijaysingh-axway commented:
For color it is working perfect in iOS. I'll check for image

ewanharris committed:
docs: correct summary

vijaysingh-axway commented:
@ewanharris  While testing for images -
1. In iOS 13, if I run app in light mode and go to setting (without killing app) and change to dark mode image is not changing. If I kill the app and run again, it get changed. I tried to add image view and for image view it is changing the image in dark mode and light mode. Probably some issue in SDK. I can check on this. I'll test for images in other component as well.
2. In iOS < 13, image is not changing because iOS < 13 does not support dark mode. Can we make a mapping similar to color.json for images as well? So that we can have it for iOS < 13 as well.
3. I guess, we need a wiki page for explaining this and an example app.

jquick-axway commented:
I think what @ewanharris is shooting for is that we should also have an "event" that gets fired when the OS' dark/light theme has changed. The theme event listener can then call the `fetchSemanticColor()` method and re-apply colors to all the views. Right?

hansemannn commented:
@jquick-axway Nope, images and colors are automatically changed by iOS if the appearance is changed - at least it does so natively. But I've added a `traitcollectionchange` event in some other PR is triggered if the appearance changes, so other changes to the app can be made as well (e.g. analytics)

ewanharris committed:
test: add tests

jquick-axway commented:
> Nope, images and colors are automatically changed by iOS if the appearance is changed

I get that, but this PR's `fetchSemanticColor()` method is documented to return a color `string` type in its `*.yml` file... which would suggest to anyone only reading our API docs that you would have to re-fetch the color string if the theme changes and re-apply it to all views. (Back when I used to do Windows desktop development, this is exactly what you had to do too.)

I can see in the `*.m` file that it returns a `TiColor` object instead of a string. So, what's the real truth here?

ewanharris commented:
So on iOS 13+, it returns a TiColor object (how do we even document that?), which will handle the dark mode setting change for a user. In every other situation, the user will get the string from the JSON object they gave us.

I totally overlooked that the UI wont refresh on anything other than iOS 13, but I'm not sure how we'd even achieve that?

@vijaysingh-axway For your point 2, are you thinking about a function that returns the correct image for a given dark mode setting? I think in theory that should just be prefixing `-dark` onto the image based on the naming conventions we have.

jquick-axway commented:
> So on iOS 13+, it returns a TiColor object (how do we even document that?), which will handle the dark mode setting change for a user.

Okay. That's fine. We just need to document this, because this is not the behavior I would have expected. Perhaps document that it returns an undocumented color `object` type?

I say this because Android and Windows do not have an equivalent. On those platforms, you set literal color values. (On Android, we could theoretically create our own custom `Drawable` class which dynamically changes color, but this is not how it's usually done on that platform... and we might run into some limitations with some of Google's view classes.)

ewanharris committed:
fix(android): define getter/setter for semanticcolortype property

Due to the way Ti.UI works on Android we cant reliably track the changes, by implementing the property with the get/set syntax we can track the property changes ourselves and reliably return the correct result

hansemannn commented:
Android Q supports dark mode as well (btw, it went to final beta today). Natively, it's managed by themes, on Titanium: Dunno. Maybe by listing to the change event and re-tint internally? I hope there is a better native way on Android.

vijaysingh-axway commented:
@ewanharris  For color we have made it easy for developers, just add a resource file semantic-color.json and use it. It will work perfect for iOS < 12. Can we make a similar file for image resources, which have attribute for dark and light similar to colors. 
In case iOS < 13, return image named for 'dark' from this for dark mode and for light mode return image name for 'light'. For iOS 13+, we do not have to worry. 
One more advantage of using this approach is that we do not have to  force developers to make image name as '-dark' extension. It will have many to one mapping for light and dark.
 e.g Let's say we have an image name image1.jpg and second image as image2.jpg and I want to use same image (darkimage.jpg) in dark mode for both. In current implementation developer have to create a copy of same image and make it image1-dark.jpg and image2-dark.jpg. But if we make mapping developer do not have to do that. Same image will work.
 @hansemannn @jquick-axway your thoughts?

@ewanharris For my point 3 in previous comment, I was say a wiki page for explanation of how developer can implement dark mode in there project.

janvennemann commented:
@vijaysingh-axway I don't think an additional file for images is necessary. IMHO adhering to a specific naming pattern is easier than having an additional mapping file to maintain which only has the benefit of being able to configure the image names. I mean how likely is it that two different light mode images have the same dark mode image?

hansemannn commented:
Same here, we should go with naming conventions to eave the development process. For us, we have 130+ unique images and no issues with the images using the `dark-` suffix.

hansemannn commented on iphone/cli/commands/_build.js:
```diff
@@ -5767,6 +5775,109 @@ iOSBuilder.prototype.copyResources = function copyResources(next) {
+									red: `${defaultRGB.r}`,
+									green: `${defaultRGB.g}`,
+									blue: `${defaultRGB.b}`,
+									alpha: '1.000'
```
Can you please add alpha handling like I've added this week?

ewanharris replied:
Added support for alpha, rather than setting per colorset, I've matched xcode and allowed setting on the color type like the below

```js
{
	"textColor": {
		"dark": {
			"color": "#ff85e2",
			"alpha": "0.500"
		},
		"light": "#ff1f1f" // defaults to alpha: 1.000
	}
}
```

hansemannn replied:
Nice - even more generic! But I would really make it a 2f to be more `opacity` like. I doubt that someone sets a `0.5 %` value.

ewanharris replied:
Updated to accept a 0-100 range in the semantic.colors.json file that the SDK will then convert to the xcode-preferred `x.xxx` format

ewanharris committed:
feat: support setting alpha per color

ewanharris commented:
@vijaysingh-axway, I agree with the others, I think promoting a convention like file naming is preferable to configuration via a json file, I do agree that it isn't as easy to transfer cross-platform as the json file though.

ewanharris committed:
refactor: allow setting alpha as a 0-100 range

vijaysingh-axway reviewed (approved):
Looks good.

ewanharris commented:
@vijaysingh-axway, on the docs I started drafting something this week which I'll aim to finish next week. I'll adapt to the repo a thttps://github.com/ewanharris/darkmode-example into something we can move to the appcelerator-developer-relations org and see about updating KitchenSink too

ewanharris committed:
fix(ios): guard in sdk 11 check

ewanharris committed:
docs: add docs

vijaysingh-axway committed:
Merge branch 'master' into TIMOB-27126

vijaysingh-axway committed:
Merge branch 'master' into TIMOB-27126

keerthi1032 commented:
FR Passed.SemanticColor is supported.fetches color and displayed based on dark mode/light mode selected on 13 device. images displayed correctly for dark and light mode.
Test Environment:
Name                        = Mac OS X
  Version                     = 10.14.5
  Architecture                = 64bit
Node.js
  Node.js Version             = 10.16.2
  npm Version                 = 6.9.0
Titanium CLI
  CLI Version                 = 5.2.1
Titanium SDK
  SDK Version                 =localSDK 8.2.0.v20190902191630
Device -iPhone XR iOS 13, iPhone X iOS 11(for iOS 11 app runs normal )
Simulator -iPhone 8 iOS 13,

keerthi1032 committed:
Merge branch 'master' into TIMOB-27126
